File: bindings/rust/src/graph/edges.rs
//! Edge operations for Graph.

use super::Graph;
use crate::utils::{rel_type_pattern, sanitize_rel_type, PropertyValue};
use crate::{CypherResult, Result, Value};

impl Graph {
    /// Check if a directed edge exists from source to target.
    pub fn has_edge(
        &self,
        source_id: &str,
        target_id: &str,
        rel_type: Option<&str>,
    ) -> Result<bool> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) RETURN count(r) AS cnt",
            rel_pattern
        );
        let result = self
            .connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        if result.is_empty() {
            return Ok(false);
        }
        let cnt: i64 = result[0].get("cnt").unwrap_or(0);
        Ok(cnt > 0)
    }

    /// Get edge properties between two nodes.
    pub fn get_edge(
        &self,
        source_id: &str,
        target_id: &str,
        rel_type: Option<&str>,
    ) -> Result<Option<Value>> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) RETURN r",
            rel_pattern
        );
        let result = self
            .connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        if result.is_empty() {
            return Ok(None);
        }
        Ok(result[0].get_value("r").cloned())
    }

    /// Create or update an edge between two nodes.
    ///
    /// If an edge of the same type already exists, its properties are updated
    /// (merge semantics — existing properties not in `props` are preserved).
    /// If no edge of that type exists, a new one is created.
    pub fn upsert_edge<I, K, V>(
        &self,
        source_id: &str,
        target_id: &str,
        props: I,
        rel_type: &str,
    ) -> Result<()>
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: Into<PropertyValue>,
    {
        let safe_rel_type = sanitize_rel_type(rel_type);

        let props: Vec<(String, PropertyValue)> = props
            .into_iter()
            .map(|(k, v)| (k.as_ref().to_string(), v.into()))
            .collect();

        let merge_query = format!(
            "MATCH (a {{id: $src}}), (b {{id: $tgt}}) MERGE (a)-[r:{}]->(b)",
            safe_rel_type
        );
        self.connection()
            .cypher_builder(&merge_query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;

        if !props.is_empty() {
            let set_parts: Vec<String> = props
                .iter()
                .map(|(k, v)| format!("r.{} = {}", k, v.to_cypher()))
                .collect();
            let set_str = set_parts.join(", ");
            let set_query = format!(
                "MATCH (a {{id: $src}})-[r:{}]->(b {{id: $tgt}}) SET {}",
                safe_rel_type, set_str
            );
            self.connection()
                .cypher_builder(&set_query)
                .param("src", source_id)
                .param("tgt", target_id)
                .run()?;
        }

        Ok(())
    }

    /// Create or update an edge identified by a caller-assigned edge id.
    ///
    /// The edge is matched/merged on `edge_id` (stored as an `id` property on
    /// the relationship) instead of on the (source, target, rel_type) triple,
    /// so multiple parallel edges can exist between the same two nodes with
    /// the same relationship type. Repeated calls with the same `edge_id`
    /// update that edge's properties in place; different `edge_id`s create
    /// distinct edges.
    ///
    /// Both source and target nodes must exist.
    pub fn upsert_edge_with_id<I, K, V>(
        &self,
        source_id: &str,
        target_id: &str,
        props: I,
        rel_type: &str,
        edge_id: &str,
    ) -> Result<()>
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: Into<PropertyValue>,
    {
        let safe_rel_type = sanitize_rel_type(rel_type);

        let props: Vec<(String, PropertyValue)> = props
            .into_iter()
            .map(|(k, v)| (k.as_ref().to_string(), v.into()))
            .collect();

        let merge_query = format!(
            "MATCH (a {{id: $src}}), (b {{id: $tgt}}) MERGE (a)-[r:{} {{id: $eid}}]->(b)",
            safe_rel_type
        );
        self.connection()
            .cypher_builder(&merge_query)
            .param("src", source_id)
            .param("tgt", target_id)
            .param("eid", edge_id)
            .run()?;

        if !props.is_empty() {
            let set_parts: Vec<String> = props
                .iter()
                .map(|(k, v)| format!("r.{} = {}", k, v.to_cypher()))
                .collect();
            let set_str = set_parts.join(", ");
            let set_query = format!(
                "MATCH (a {{id: $src}})-[r:{} {{id: $eid}}]->(b {{id: $tgt}}) SET {}",
                safe_rel_type, set_str
            );
            self.connection()
                .cypher_builder(&set_query)
                .param("src", source_id)
                .param("tgt", target_id)
                .param("eid", edge_id)
                .run()?;
        }

        Ok(())
    }

    /// Delete the directed edge between two nodes.
    pub fn delete_edge(
        &self,
        source_id: &str,
        target_id: &str,
        rel_type: Option<&str>,
    ) -> Result<()> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) DELETE r",
            rel_pattern
        );
        self.connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        Ok(())
    }

    /// Get all edges in the graph.
    pub fn get_all_edges(&self) -> Result<CypherResult> {
        self.connection()
            .cypher("MATCH (a)-[r]->(b) RETURN a.id AS source, b.id AS target, r")
    }
}

File: bindings/typescript/src/graph/edges.ts
// Edge read/delete/upsert operations as pure functions.
//
// Each takes a Connection first so Graph can delegate in three lines (see
// graph/index.ts). Mirrors bindings/python/src/graphqlite/graph/edges.py.
import type { Connection } from '../connection.ts';
import { sanitizeRelType, assertIdentifier } from '../utils.ts';
import type { CypherRow, CypherValue } from '../result.ts';

/**
 * The `[r{...}]` relationship pattern. When a `relType` is given it is sanitized
 * (never validated/thrown — {@link sanitizeRelType} coerces to a safe token) and
 * interpolated as `:TYPE`; otherwise the pattern is bare `[r]`. Node ids are
 * always bound (`$src`/`$tgt`), never interpolated.
 */
function relPattern(relType?: string): string {
  return relType ? `:${sanitizeRelType(relType)}` : '';
}

/**
 * Whether an edge exists between two nodes (optionally of a given type). Parses
 * like {@link import('./nodes.ts').hasNode}: empty → false, falsy `cnt` → false,
 * else `Number(cnt) > 0` (mirrors edges.py:31-34).
 */
export function hasEdge(
  conn: Connection,
  sourceId: string,
  targetId: string,
  relType?: string,
): boolean {
  const result = conn.cypher(
    `MATCH (a {id: $src})-[r${relPattern(relType)}]->(b {id: $tgt}) RETURN count(r) AS cnt`,
    { src: sourceId, tgt: targetId },
  );
  if (result.length === 0) {
    return false;
  }
  const cnt = result[0]?.['cnt'];
  return cnt ? Number(cnt) > 0 : false;
}

/** Fetch the edge between two nodes, or `null` if none. `r` is returned unmodified. */
export function getEdge(
  conn: Connection,
  sourceId: string,
  targetId: string,
  relType?: string,
): CypherValue | null {
  const result = conn.cypher(
    `MATCH (a {id: $src})-[r${relPattern(relType)}]->(b {id: $tgt}) RETURN r`,
    { src: sourceId, tgt: targetId },
  );
  if (result.length === 0) {
    return null;
  }
  return result[0]?.['r'] ?? null;
}

/** Delete the edge(s) between two nodes (optionally of a given type). */
export function deleteEdge(
  conn: Connection,
  sourceId: string,
  targetId: string,
  relType?: string,
): void {
  conn.cypher(`MATCH (a {id: $src})-[r${relPattern(relType)}]->(b {id: $tgt}) DELETE r`, {
    src: sourceId,
    tgt: targetId,
  });
}

/**
 * All edges as `{ source, target, r }` rows. The `toList()` result is returned
 * **unmodified** (mirrors edges.py:145-148); callers depend on those exact keys.
 */
export function getAllEdges(conn: Connection): CypherRow[] {
  const result = conn.cypher('MATCH (a)-[r]->(b) RETURN a.id AS source, b.id AS target, r');
  return result.toList();
}

/**
 * Create or update an edge via `MERGE` (mirrors edges.py:58-119). Unlike
 * `upsertNode`, this never calls `hasEdge`/`hasNode` — MERGE is idempotent.
 *
 * - **Step 1 (MERGE)** matches both nodes and merges the relationship. Without
 *   `edgeId` the merge key is `(source, target, relType)`; with `edgeId` it is
 *   the caller-assigned `id` property (parallel edges of the same type).
 * - **Step 2 (SET)** runs only when `edgeData` is non-empty, as a **single**
 *   query — values bound (`$v0`, `$v1`, …), property keys interpolated (contrast
 *   `upsertNode`, which issues one query per key).
 *
 * If either node is missing the MATCH yields nothing and the whole thing is a
 * silent no-op — no exception. `relType` defaults to `"RELATED"` and is
 * sanitized; every property key is validated with {@link assertIdentifier}.
 */
export function upsertEdge(
  conn: Connection,
  sourceId: string,
  targetId: string,
  edgeData: Record<string, unknown>,
  relType: string = 'RELATED',
  edgeId?: string,
): void {
  const safe = sanitizeRelType(relType);
  const keys = Object.keys(edgeData);
  // Validate keys before any write so a bad key never leaves a dangling edge.
  for (const key of keys) {
    assertIdentifier(key, 'property');
  }

  let relMatch: string;
  let baseParams: Record<string, unknown>;
  if (edgeId === undefined) {
    conn.cypher(`MATCH (a {id: $src}), (b {id: $tgt}) MERGE (a)-[r:${safe}]->(b)`, {
      src: sourceId,
      tgt: targetId,
    });
    relMatch = `[r:${safe}]`;
    baseParams = { src: sourceId, tgt: targetId };
  } else {
    conn.cypher(`MATCH (a {id: $src}), (b {id: $tgt}) MERGE (a)-[r:${safe} {id: $eid}]->(b)`, {
      src: sourceId,
      tgt: targetId,
      eid: edgeId,
    });
    relMatch = `[r:${safe} {id: $eid}]`;
    baseParams = { src: sourceId, tgt: targetId, eid: edgeId };
  }

  if (keys.length > 0) {
    const params: Record<string, unknown> = { ...baseParams };
    const setParts = keys.map((key, i) => {
      const paramName = `v${i}`;
      params[paramName] = edgeData[key];
      return `r.${key} = $${paramName}`;
    });
    conn.cypher(
      `MATCH (a {id: $src})-${relMatch}->(b {id: $tgt}) SET ${setParts.join(', ')}`,
      params,
    );
  }
}
